### ISSUE TITLE
Bug: no-misleading-character-class triggers on seemingly-OK code

### ISSUE THREAD

**Issue by domenic** [eslint/eslint#15080]
### Environment

Node version: 16.9.1
npm version: 7.21.1
Local ESLint version: 7.32.0
Global ESLint version: (none)
Operating System: Windows 10

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

Just using eslint:recommended. This can be [demoed on eslint.org](https://eslint.org/demo#eyJ0ZXh0IjoiL1tcXHUwMEI3XFx1MDMwMC1cXHUwMzZGXS91LnRlc3QoXCJcIik7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUtZWxzZS1pZiI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHkiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW1wb3J0LWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1pc2xlYWRpbmctY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOjIsIm5vLW5ldy1zeW1ib2wiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXByb3RvdHlwZS1idWlsdGlucyI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zZXR0ZXItcmV0dXJuIjoyLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZGVmIjoyLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6Miwibm8tdW5yZWFjaGFibGUiOjIsIm5vLXVuc2FmZS1maW5hbGx5IjoyLCJuby11bnNhZmUtbmVnYXRpb24iOjIsIm5vLXVudXNlZC1sYWJlbHMiOjIsIm5vLXVudXNlZC12YXJzIjoyLCJuby11c2VsZXNzLWNhdGNoIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6Miwibm8td2l0aCI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjJ9LCJlbnYiOnt9fX0=)
</details>

```js
/[\u00B7\u0300-\u036F]/u.test("");
```

### What did you expect to happen?

No error, if I am understanding https://eslint.org/docs/rules/no-misleading-character-class correctly. Those docs imply the error will trigger when I include in my source text a multi-code-point character, which I have not done; I've only used Unicode escapes.

### What actually happened?

```
4:3   error  Unexpected combined character in character class  no-misleading-character-class
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

For context, I am trying to reproduce https://www.w3.org/TR/REC-xml/#NT-NameStartChar as a RegExp. The `#xB7 | [#x0300-#x036F]` part of the grammar is what maps to my `[\u00B7\u0300-\u036F]` character class.

It's possible that this is a doc issue, as I'm actually doing something dangerous or unusual, and so the error should trigger? But if I am, it's not clear at all from the docs what I'm doing that's bad.

**Comment by mdjermanovic** [comment]
Hi @domenic, thanks for the issue!

By test cases for this rule, this seems to be intended behavior:

https://github.com/eslint/eslint/blob/47be8003d700bc0606495ae42610eaba94e639c5/tests/lib/rules/no-misleading-character-class.js#L97-L100

I couldn't find any discussions about this in the original issue https://github.com/eslint/eslint/issues/10049 and PR https://github.com/eslint/eslint/pull/10511, but I think that reporting this character class makes sense - `[\u0041\u0301]` might have been intended to match with `"Á"`, the user is aware that the targeted combined character needs two code units, but maybe unaware that this doesn't match as a single character, unlike surrogate pairs.

In your `/[\u00B7\u0300-\u036F]/u` example, concern is that the intent might be to search for `"\u00B7\u0300"`, ... `"\u00B7\u036F"` instead of any of `"\u00B7"`, `"\u0300"`, ... `"\u036F"`.

The rule does not warn if combining characters are not preceded by other characters in same `[]`, as it's clear that the intent is to match those characters separately:

```js
/[\u00B7\u0300-\u036F]/u.test(""); // error

/[\u0300-\u036F\u00B7]/u.test(""); // ok

/[\u0300-\u036F]|\u00B7/u.test(""); // ok

/\u00B7|[\u0300-\u036F]/u.test(""); // ok

/[\u0300-\u036F]|[\u00B7]/u.test(""); // ok

/[\u00B7]|[\u0300-\u036F]/u.test(""); // ok
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiL1tcXHUwMEI3XFx1MDMwMC1cXHUwMzZGXS91LnRlc3QoXCJcIik7IC8vIGVycm9yXG5cbi9bXFx1MDMwMC1cXHUwMzZGXFx1MDBCN10vdS50ZXN0KFwiXCIpOyAvLyBva1xuXG4vW1xcdTAzMDAtXFx1MDM2Rl18XFx1MDBCNy91LnRlc3QoXCJcIik7IC8vIG9rXG5cbi9cXHUwMEI3fFtcXHUwMzAwLVxcdTAzNkZdL3UudGVzdChcIlwiKTsgLy8gb2tcblxuL1tcXHUwMzAwLVxcdTAzNkZdfFtcXHUwMEI3XS91LnRlc3QoXCJcIik7IC8vIG9rXG5cbi9bXFx1MDBCN118W1xcdTAzMDAtXFx1MDM2Rl0vdS50ZXN0KFwiXCIpOyAvLyBvayIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWVsc2UtaWYiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWltcG9ydC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1zZWxmLWFzc2lnbiI6Miwibm8tc2V0dGVyLXJldHVybiI6Miwibm8tc2hhZG93LXJlc3RyaWN0ZWQtbmFtZXMiOjIsIm5vLXNwYXJzZS1hcnJheXMiOjIsIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjoyLCJuby11bmRlZiI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsIm5vLXdpdGgiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7fX19)

In my opinion, this behavior is correct, but we should certainly improve documentation for this rule.

**Comment by domenic** [comment]
Thanks for the research! Hmm.

It would be nice if there were a version of this rule that did what the documentation currently says, which is prevent actually-misleading cases where the result appears as a signal grapheme. That is, taking an example from the doc: I think `/^[❇️]$/u` is indeed misleading because it looks like one character. But `/^[\u2747\uFE0F]$/u` is, in my opinion, not misleading, because it's pretty clear that there are two code points. (This is despite the fact that they are equivalent semantically.)

In other words, it would be good to have a mode for this rule that exempts escape sequences.

**Comment by mdjermanovic** [comment]
> But `/^[\u2747\uFE0F]$/u` is, in my opinion, not misleading, because it's pretty clear that there are two possible code points, of which only one will be matched. (This is despite the fact that the two regexps are equivalent semantically.)
> 
> In other words, it would be good to have a mode for this rule that exempts escape sequences.

I'm not opposed to adding another mode, but I still think that `[\u2747\uFE0F]` can be misleading as the author may have thought it was a single code point because it appears as a single character, like `[\uD83D\uDC4D]` for example.

```js
/ab[c\uD83D\uDC4D]/u.test("ab👍"); // `true`, this works
/ab[c\uD83D\uDC4D]/u.exec("ab👍")[0]; // 'ab👍', as expected
/ab[c\uD83D\uDC4D]/u.exec("ab👍")[0].length; // 4

/ab[c\u2747\uFE0F]/u.test("ab❇️"); // `true`, this must be working too
/ab[c\u2747\uFE0F]/u.exec("ab❇️")[0]; // 'ab❇', looks good
/ab[c\u2747\uFE0F]/u.exec("ab❇️")[0].length; // 3! the character class was misleading 
```

@eslint/eslint-tsc thoughts about this? I think the current behavior makes sense as default behavior, but we can consider adding an option to only warn when the targeted sequences appear literally as source code characters. This seems doable for regex literals, but may be problematic for `RegExp("...")`.

**Comment by btmills** [comment]
Thanks both for the thorough explanations so far. I'm not opposed to a mode that would allow explicit code point escapes. Were this an opt-in rule, I'd say such a mode should be disabled by default for maximum safety. This being included in `eslint:recommended` gives me slight pause because we try hard to avoid any false positives in `eslint:recommended`, and this could be seen as a false positive. I'm still leaning toward leaving the current behavior as the default because a mode to allow this feels like a power user feature where those who need it will find it, but I wanted to call that out to see what others think.

**Comment by nzakas** [comment]
> leaving the current behavior as the default because a mode to allow this feels like a power user feature where those who need it will find it

💯

**Comment by domenic** [comment]
Thanks so much all for being willing to consider a mode. My two cents:

- Personally it's hard to imagine that if someone wrote `/[\u00B7\u0300-\u036F]/u` they expect anything different than two code points being matched. That is, I cannot personally imagine when that code would be a bug and thus deserving of being warned about by ESLint.

- However, I also agree that matching code points in this way is a power user feature. So if power users have to do an extra step to get the sensible behavior, that seems totally fine. Especially since at least some folks (per https://github.com/eslint/eslint/issues/15080#issuecomment-922551819) disagree with my assessment of the potential bugginess.

**Comment by nzakas** [comment]
Marking as accepted because there is agreement on adding a new option for this.

**Comment by fasttime** [comment]
I can work on this when #18082 is settled.

**Comment by nzakas** [comment]
@fasttime #18082 has been merged. Are you up for looking at this?

**Comment by fasttime** [comment]
> @fasttime #18082 has been merged. Are you up for looking at this?

Sure! This should be doable now that we have the tooling to map characters back to source code. I'll be able to focus on this next week.

**Comment by nzakas** [comment]
Great, thanks!
